prime-router/docs/proposals/0001-authentication.md

```diff
@@ -26,7 +29,93 @@ The inspiration to use the combination of Private Key JWT and hosted JWKS came f
+The new auth info will include:
+- The Sender's JSON Web Key Set (JWK Set)
+- (for a Phase 2:  Sender's Public Key URL)
+- Note: we will use the Sender's fullName (aka orgName.senderName) as the unique client_id, per the FHIR spec.
+
+We'll need to write a document for our Senders, explaining how to generate a public/private key, and how to keep the private key secure.   We'll need a process for rotating keys, if the Sender is not using the Public Key URL.   Since we have relatively few senders, I think there's not a high priority on automating that right now.  We will need to track contact information of the person at the Sender who is responsible for the keypair.  This may not be the same person who handles the actual test data/results.
+
+The sender can give us multiple keys in the JWK Set
+
+In this initial release, we'll require submission to the settings API as properly formatted JWK Set.   However, in the future, we may want to just accept a base64 encoded public key submission, and do the conversion to a JWK as a convenience.
+
+### Step Two: Signature Verification Step: Sender sends signed JWT.  If valid, response is the ReportStream token.
+
+We'll need to document for our Senders how to generate the 'Sender Token', the signed JWT they send to us, generated using their private key.
+
+The Sender Token will be sent as a URL parameter, not a header. (? I'd rather do it as a header, but I think Rick prefers parameters)
+
+We'll support both RS384 and ES384 web signature algorithms.
+
+(Phase Two Note):  The sender can also send the URL with the signed JWT, as part of the request.  Per the FHIR protocol, we must check that against the URL originally provided at the time of PreAuth.  If it differs, we reject.
+
+(Phase Two Note): How will we handle cache-control headers? Ans:  Per FHIR protocol, we should follow timeline in any cache-control request from the sender.
+
+We will 
+1. lookup and validate the client ID
+2  validate the JWT signature against the public key we have stored in the settings table for that client_id, Key ID (`kid`).  We'll use the [JJWT libraries](https://github.com/jwtk/jjwt) to do JWT verification. (?)
+2. check that the JWT has not been previously encountered within the max JWT lifetime (5 minutes)
+
+We will then pull our ReportStream secret from our vault, and use it to create and sign a ReportStream token, for use by the Sender.   FHIR calls for a 5 minute end of life.
+
+### Step Three: Actual API Usage :  Sender uses the token to submit data to `api/reports`
+
+The current 'FUNCTION' level AuthorizationLevel will be changed to 'ANONYMOUS'. (See https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=csharp#secure-an-http-endpoint-in-production)
+
+The ReportFunction will check the ReportStream token upon entry to the function code.  That is, we will not implement a 'gateway' function in front or api/reports.   (And, as mentioned earlier, no dependency on Okta)
+
```

Comment:
Maybe worth discussing how the Report function verifies the token on entry. Does the Report function have a shared secret with the Auth function? I think so. Where is this secret stored? In the Azure Key Vault.